src/printf/printf.h

```diff
@@ -86,7 +102,7 @@ void putchar_(char character);
  * @param format A string that specifies the format of the output
  * @return The number of characters that are written into the array, not counting the terminating null character
  */
-int printf_(const char* format, ...) ATTR_PRINTF(1, 2);
+printf_int_t printf_(const char* format, ...) ATTR_PRINTF(1, 2);
```

Comment:
This is not possible. The printf() functions must be compatible with the standard library. If the MISRA standard is incompatible with using the standard library, then we have a problem.

Reply:
This is not necessarily incompatible with the std lib. If you're not defining PRINTF_CUSTOM_INT_xx then ```printf_int_t ``` is a plain int. For MISRA compatibility you'd define PRINTF_CUSTOM_INT_32 to make ```printf_int_t ``` an ```int32_t```.

Reply:
Ah, ok. I guess we can do something like that. But - I still have to see the literal int in the signature. So I would rather have  something like:
```
#ifndef PRINTF_CUSTOM_INT_xx
int printf_(const char* format, ...) ATTR_PRINTF(1, 2);
#else
printf_int_t printf_(const char* format, ...) ATTR_PRINTF(1, 2);
#endif
```
or maybe even:
```
#ifndef PRINTF_CUSTOM_INT_xx
int def1
int def2
int def3
#else 
printf_int_t def1
printf_int_t def2
printf_int_t def3
#endif
```

Reply:
I've update the header file. Is this better?

Reply:
Also... what does MISRA have to say about the standard library version of these functions? I mean, they return a plain int, after all.

Reply:
Actually, I've just read the MISRA rule regarding typedefs. It says:

> 6.3 (adv) typedefs that indicate size and signedness should be used in place of the basic numerical types.

So, first of all, it's a "should", not a "shall" or "will". But even more importantly - the changes you propose don't bring you much closer to satisfying this rule, since the typedef you proposed indicates neither size nor signedness. I believe this advisory is about preferring `int32_t`, `uint8_t` and such over `int`, `unsigned`, `char` and such.

Reply:
Ah ok. I'll see about that new pull request.

Basically MISRA indicates (among others) to avoid the standard library methods regarding time/date, character handling, stream operators and allocation/deallocation. It more or less recognizes one cannot do without, but discourages dependence.

More to the point, it says "For example, the essential type rules apply to the types of the arguments passed to functions specified in the standard library and to their results."

Reply:
> Actually, I've just read the MISRA rule regarding typedefs. It says:
> 
> > 6.3 (adv) typedefs that indicate size and signedness should be used in place of the basic numerical types.
> 
> So, first of all, it's a "should", not a "shall" or "will". But even more importantly - the changes you propose don't bring you much closer to satisfying this rule, since the typedef you proposed indicates neither size nor signedness. I believe this advisory is about preferring `int32_t`, `uint8_t` and such over `int`, `unsigned`, `char` and such.
That's what I'm doing by defining printf_int_t as int16_t , int32_t or int64_t. This way the return value uses a type with defined signedness and size.

Reply:
> Basically MISRA indicates (among others) to avoid the standard library methods regarding time/date, character handling, stream operators and allocation/deallocation. It more or less recognizes one cannot do without, but discourages dependence.

Well, this library is about providing these same problematic functions. But - we'll see how it looks in the new PR. Closing this one for now.